### ISSUE TITLE
[Proposal] Include env var to disable kernel memory accounting when kernel <= 3.10

### ISSUE THREAD

**Issue by ryarnyah** [opencontainers/runc#1898]
When using RedHat or CentOS, base kernel is 3.10 (with some backport and CONFIG_MEMCG_KMEM..).
We encounter on my current project the same error as described is attached issues.

To avoid reboot all ours nodes in production every week (or recompile kernel on every update) we propose to disable it with an env var `DISABLE_KERNEL_MEMORY_ACCOUNTING` in `libcontainer/cgroups/fs/memory.go`.

See: https://github.com/ryarnyah/runc/tree/1898-add-optional-kernel-accounting

See: https://github.com/scality/kubernetes/commit/b04b0506e49b1b60ea7c8b74a5ca2edbf341cd6c
See: https://github.com/kubernetes/kubernetes/issues/61937
See: https://github.com/opencontainers/runc/pull/1350
See: https://github.com/moby/moby/issues/29638

**Comment by cyphar** [comment]
If you remove all kernel memory accounting options from `config.json` then `runc` shouldn't be touching kernel memory. If we are touching kernel memory for a good reason, we should not touch it if we get `ENOENT` (I don't think we have code for that at the moment -- but it is how I would implement it).

NACK on using an environment variable for this. It feels like a very hacky way of solving the problem, leads to OCI compliance issues (now `runc` will ignore kernel memory settings entirely if the env-flag is set), requires users to manually set the environment variable, and is just hiding a deeper problem.

Auto-detecting by seeing if we get `ENOENT` would be a far better way of solving it (because presumably users would only pass `RUNC_DISABLE_KERNEL_MEMORY_LIMIT` if kernel memory limits aren't exposed by their kernel -- which is something we can detect without asking the user to tell us about it).

**Comment by cyphar** [comment]
Sorry, I misunderstood what the issue was (it's been a while since the last `kmem` bugs were fixed). So there's a kernel bug in pre-4.0 kernels that's causing setting `memory.kmem` to result in a kernel memory leak. Obviously this is very bad (and RedHat should probably work on fixing it -- given that it is a problem in their kernels as far as I understand).

The wish for an environment variable or a flag makes some more sense now, but I'm still not really thrilled about it (we added `--no-new-keyring` for similar reasons and I'm still not a huge fan of it). Kernel version tests aren't really very useful because distributions do backports, so those aren't an option.

From (2-year-old) memory, the reason why we always enable `kmem` accounting is because you have to write _some_ value in order to be able to update it later. This is a fairly frustrating restriction of `kmem` (though I believe it is a performance-related restriction) and it isn't clear how exactly we can disable it without also breaking `runc update` (the reason we made this change in the first place -- as @hqhq has said in the linked issues).

tl;dr: Not sure what the best way forward is...

**Comment by crosbymichael** [comment]
Ya, lets not use flags or env vars to resolve this issue.

**Comment by kolyshkin** [comment]
Alternative proposal, checking if the kernel is a one from RHEL7: https://github.com/opencontainers/runc/pull/1920

**Comment by ryarnyah** [comment]
Closed in favor of https://github.com/opencontainers/runc/pull/1921
